teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -0,0 +1,213 @@
+etcd over time. See [Relevant Issues under Motivation](#motivation).
+
+## Requirements
+
+- Run metadata should be available even after deletion from the Pipeline
+  controller data plane.
+- Runs should be deleted as soon as it can be done safely (i.e. we have
+  confirmed the data is stored durably elsewhere).
+- Deletion should occur with some configurable delay to allow other integrations
+  watching Runs to respond to events.
+- Operators should be able to opt in / out to deletion behavior.
+
+## Proposal
+
+### Delete Runs when complete
+
+The
+[Tekton Results watcher](https://github.com/tektoncd/results/blob/main/docs/watcher/README.md)
+already watches for changes to TaskRun and PipelineRuns to upload Records for
+long term storage. Once we have verified that 1)
+[the Run has completed](https://github.com/tektoncd/pipeline/blob/master/docs/runs.md#monitoring-execution-status)
+and 2) the Run Record has been successfully stored, we know we can delete the
+Run without affecting execution.
+
+### Grace Period
+
+Tekton Results may not be the only controller listening to changes in TaskRun
+and PipelineRun CRDs (e.g. Notifications that happen post-Run execution).
+Because of this, deleting the Run immediately might cause unintended race
+conditions and affect other integrations using a similar mechanism.
+
+To address this, we should have a user configurable grace period for Run
+deletion, and require that the time elapsed since the Record's last update time
+(since the k8s ObjectMeta does not have a similar concept of last update time)
+is greater than the configured grace period before deletion. If the grace period
+has not been met yet, then we should simply
+[re-enqueue](https://pkg.go.dev/knative.dev/pkg@v0.0.0-20210208175252-a02dcff9ee26/controller#Impl.EnqueueKeyAfter)
+the Run to be handled by the controller at a later point in time.
+
+To start, this configuration will be set by the operator of the Results Watcher via
+a flag. We may consider adding additional end user controls for this behavior
+later (see [Future Work](#grace-period-api-configuration)), but this is out of
+scope for this initial work.
+
+### Notes/Caveats (optional)
+
+#### Future Work
+
+##### Grace Period API Configuration
+
+It might be useful to allow users to configure the grace period via the API
+(either on a per namespace or per resource level). This is a backwards
+compatible improvement that can be added later, so we are considering this to be
+out of scope of this design.
+
+### Risks and Mitigations
+
+#### Early Deletion
+
+Because it's impossible to know when every third-party integration that watches
+Runs is complete, we risk deleting a Run before another system can process it.
+To prevent this, we will rely on a grace period to minimize this risk.
+Integrations that require access to Run metadata for long periods of time after
+completion (O(minutes)), should consider using the Results API as the source of
```

Comment:
i could imagine someone wanting this grace period to be something like a day (or even a week or more) - does this solution also support configuring the grace period to be on the order of hours or even days?

as a possible reason, i could imagine folks might be using other tools that might integrate with the k8s API and it might not be reasonable to expect they can update those tools to use the Result API as the source of truth

Reply:
> so don't try to schedule deletion 300 years in the future and we should be good 

lolololol

i wonder how big that queue can get 🤔  anyway sounds like we're covered :D